Update to latest version of dom-anchor-text-{quote, position} libs

robertknight opened the pull request:
This updates the text position and quote anchoring libraries to the latest major version which fixes several failing anchoring tests.
- Update `types.coffee` to use the modified interfaces of the libraries
  which now export functions rather than classes.
- Update test expectations for the HTML anchoring tests as the updated
  dom-anchor-text-position library fixes a bug that caused most of the
  failures

robertknight commented:
Marking as WIP for the moment since I'd like to validate that there have been no unexpected changes to the selectors that are captured first.

robertknight commented:
This is blocked by https://github.com/hypothesis/client/pull/179 and should be rebased after that merges.

robertknight commented on h/static/scripts/annotator/anchoring/types.coffee:
```diff
@@ -47,7 +49,62 @@ class RangeAnchor
+###*
+# Converts between TextPositionSelector selectors and Range objects.
+###
+class TextPositionAnchor
```
`html.coffee` assumes that Range <-> selector conversion for all 4 types of selector has the same interface, which looks something like this:

```ts
interface AnchorFactory {
  fromRange(root: Element, range: Range, options: Object): Anchor
  fromSelector(root: Element, selector: Selector): Anchor
}

interface Anchor {
  toSelector(): Selector;
  toRange(options: Object): Range;
}

interface Selector {
  type: 'TextQuoteSelector' | 'TextPositionSelector' | 'FragmentSelector' | 'RangeSelector';
  // Selector-specific fields...
}
```

The `TextPositionAnchor` and `TextQuoteAnchor` classes added here are wrappers around the interface exported by the latest version of the quote and position anchoring libraries so that this remains the case.

The other way to handle this would be to adopt the new interface of the `dom-anchor-text-{quote, position}` libraries as the interface for all 4 types of selector. I think that will be cleaner overall but I would like to explore that as a follow-up refactoring.

robertknight commented:
For reference https://github.com/tilgovi/dom-anchor-text-position/commit/d0d11e3630588e4f8b2dbddf5cc70d38b2491677 is the main upstream change which I believe fixes most or all of the failing test cases.

robertknight commented:
> Have you checked that this version of the code correctly anchors existing annotations on heavily annotated pages from the current service?

I'm currently doing some extended testing of this on a few different pages.

robertknight commented:
I have done some tests with a few [heavily annotated pages](https://hypothes.is/groups/wqvn7Lar/hypothesis-reading?q=tag%3Astress-tests) using a prod extension build in Chrome. All looks good. I'm now looking into giving these a spin in other browsers.

robertknight commented:
Bah, I found a regression in Edge:

**Steps to reproduce**

1. Run `gulp watch` in the client repo
2. Navigate to http://localhost:3000 and select the text "Screenshot of Hypothesis client"
3. Click Annotate

**Expected** - The selected text is highlighted yellow
**Actual** - A section partially to the right of the selected text is highlighted

![edge-range-select](https://cloud.githubusercontent.com/assets/2458/21357010/9e9b6518-c6cb-11e6-9bad-46f440eac5b6.png)

![edge-highlight](https://cloud.githubusercontent.com/assets/2458/21357015/a2911fbe-c6cb-11e6-8b14-30fd81112403.png)

The same steps result in a correct highlight in Firefox and Chrome.

robertknight commented:
It looks like there is a difference in behavior of `Range.prototype.toString` (used to convert a range to start/end positions within the text of the root node) and `Node.textContent` (used to perform the opposite conversion) in Edge wrt. new lines.

In the test case above, the value of `range.toString()` includes `\r\n` new lines whereas the value of `root.textContent` uses only `\n` for new lines. What happens when describing the range is:

1. dom-anchor-text-quote calls dom-anchor-text-position's `fromRange()` function which uses `range.toString().length` and `prefix.toString().length` (where prefix is a range spanning from the start of the body up to the start of the range) to compute the start/end offsets of the range
2. dom-anchor-text-quote then uses  `root.textContent.substr(start, end-start)` to get the quote

Because the two methods of getting the text return different results, the resulting quote is not what the user selected.

robertknight commented:
In browsers except IE and Edge everything looks good. Unfortunately in IE/Edge this is blocked on https://github.com/tilgovi/dom-anchor-text-position/issues/4 .

tilgovi commented:
If it's helpful, I can release a version of dom-anchor-fragment that mirrors the new, function-oriented APIs. That'd be super easy.

robertknight commented:
> If it's helpful, I can release a version of dom-anchor-fragment that mirrors the new, function-oriented APIs. That'd be super easy.

It isn't urgent but that would be helpful.

robertknight commented:
Status update: The issue affecting Edge/IE has been fixed upstream in dom-anchor-text-position v4. dom-anchor-text-quote depends on dom-anchor-text-position so will need updating as well to incorporate the fix.

Once that is done, this PR will need:
1. Updating to use the newer versions of these two libs
2. Be modified to handle the fact that dom-anchor-text-quote's toRange now returns null rather than throwing an exception if the quote is not found.

tilgovi commented:
I just pushed dom-anchor-text-quote 4.0.1 to bump the dom-anchor-text-position dependency. Thanks for catching that.

robertknight commented:
I have updated this branch with a fix for the IE/Edge issue above and also added some additional tests to check behavior when anchoring with each of the the different selector types fails.
